contracts/src/utils/structs/bitmap.rs

```diff
@@ -0,0 +1,140 @@
+//! consume an entire slot for a single value.
+//!
+//! This results in gas savings in two ways:
+//!
+//! - Setting a zero value to non-zero only once every 256 times
+//! - Accessing the same warm slot for every 256 _sequential_ indices
+use alloy_primitives::U256;
+use stylus_proc::sol_storage;
+
+sol_storage! {
+    /// State of bit map.
+    pub struct BitMap {
+        /// Inner laying mapping.
+        mapping(uint256 => uint256) _data;
+    }
+}
+
+impl BitMap {
+    /// Returns whether the bit at `index` is set.
+    ///
+    /// # Arguments
+    ///
+    /// * `index` - index of boolean value at the bit map.
+    #[must_use]
+    pub fn get(&self, index: U256) -> bool {
+        let bucket = index >> 8;
+        let mask = U256::from(1) << (index & U256::from(0xff));
+        let value = self._data.get(bucket);
+        (value & mask) != U256::ZERO
+    }
+
+    /// Sets the bit at `index` to the boolean `value`.
+    ///
+    /// # Arguments
+    ///
+    /// * `index` - index of boolean value at the bit map.
+    /// * `value` - boolean value to set into the bit map.
+    pub fn set_to(&mut self, index: U256, value: bool) {
+        if value {
+            self.set(index);
+        } else {
+            self.unset(index);
+        }
+    }
+
+    /// Sets the bit at `index`.
+    ///
+    /// # Arguments
+    ///
+    /// * `index` - index of boolean value that should be set `true`.
+    pub fn set(&mut self, index: U256) {
+        let bucket = Self::get_bucket(index);
+        let mask = Self::get_mask(index);
+        let mut value = self._data.setter(bucket);
+        let prev = value.get();
+        value.set(prev | mask);
+    }
+
+    /// Unsets the bit at `index`.
+    ///
+    /// # Arguments
+    ///
+    /// * `index` - index of boolean value that should be set `false`.
+    pub fn unset(&mut self, index: U256) {
+        let bucket = Self::get_bucket(index);
+        let mask = Self::get_mask(index);
+        let mut value = self._data.setter(bucket);
+        let prev = value.get();
+        value.set(prev & !mask);
+    }
+
+    /// Get mask of value in the bucket.
+    fn get_mask(index: U256) -> U256 {
+        U256::from(1) << (index & U256::from(0xff))
+    }
+
+    /// Get bucket index.
+    fn get_bucket(index: U256) -> U256 {
+        index >> 8
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
+    use alloy_primitives::{
+        private::proptest::{
+            prelude::{Arbitrary, ProptestConfig},
+            proptest,
+        },
+        U256,
+    };
+    use stylus_sdk::{prelude::*, storage::StorageMap};
+
+    use crate::utils::structs::bitmap::BitMap;
+
+    impl Default for BitMap {
+        fn default() -> Self {
+            let root = U256::ZERO;
+            BitMap { _data: unsafe { StorageMap::new(root, 0) } }
+        }
+    }
+
+    #[motsu::test]
+    fn set_value() {
+        proptest!(|(value: U256)| {
+            let mut bit_map = BitMap::default();
```

Comment:
This one seems a bit debatable decision, but now `#[motsu::test]` macro locks storage for a test execution even if there is no contract argument.
Assuming that rust contract struct can be created during test.
And the state of this contract will be single per test.
What do you think @alexfertel ? Or may be use `with_context` function?

Reply:
I think `with_context` is slightly cleaner because we are instantiating the contract twice this way (with `default`), right? 

In any case, I'm fine either way, I'll leave it to your judgment!